Bootloader/bootloader_core.c

```diff
@@ -0,0 +1,514 @@
+}
+#endif
+
+#ifdef BOOTLOADER_CONFIG
+/******************************************************************************
+ * @brief Save the current page when BIN_END command is received 
+ * @param None 
+ * @return None
+ ******************************************************************************/
+void LuosBootloader_SaveLastData(void)
+{
+    LuosHAL_ProgramFlash(flash_addr, (uint16_t)BUFFER_SIZE, data_buff);
+}
+#endif
+
+#ifdef BOOTLOADER_CONFIG
+/******************************************************************************
+ * @brief compute crc 8 for each data
+ * @param data pointer, data len 
+ * @return crc
+ ******************************************************************************/
+void crc8(const uint8_t *data, uint8_t *crc, uint16_t polynome)
+{
+    uint16_t dbyte = 0;
+    uint16_t mix   = 0;
+
+    // zero padding data
+    dbyte = (uint16_t)*data << 8;
+    // add initial value
+    dbyte ^= (uint16_t)*crc << 8;
+    // right shift polynomial key
+    polynome = polynome << 8;
+    // shift and xor if necessary
+    for (uint8_t j = 0; j < 8; ++j)
+    {
+        mix   = dbyte & 0x8000;
+        dbyte = (dbyte << 1);
+        if (mix)
+            dbyte ^= polynome;
+    }
+
+    *crc = (uint8_t)(dbyte >> 8);
+}
+#endif
+
+#ifdef BOOTLOADER_CONFIG
+/******************************************************************************
+ * @brief compute crc for the whole binary
+ * @param data pointer, data len 
+ * @return crc
+ ******************************************************************************/
+uint8_t compute_crc(void)
+{
+    uint8_t crc   = 0x00; // initial crc value
+    uint8_t data  = 0x00;
+    uint16_t poly = 0x0007;
+
+    uint32_t data_counter  = 0;
+    uint8_t data_index     = 0;
+    uint32_t *data_address = (uint32_t *)APP_ADDRESS;
+    uint32_t data_flash    = 0;
+
+    while (data_counter < nb_bytes)
+    {
+        // read 1 word in flash memory
+        data_flash = *data_address;
+        // read 4 bytes in flash memory
+        for (data_index = 0; data_index < 4; data_index++)
+        {
+            if (data_counter < nb_bytes)
+            {
+                data = (uint8_t)(data_flash >> (8 * data_index));
+                // compute crc
+                crc8(&data, &crc, poly);
+                data_counter += 1;
+            }
+        }
+        // update data address
+        data_address += 1;
+    }
+
+    return crc;
+}
+#endif
+
+#ifdef BOOTLOADER_CONFIG
+/******************************************************************************
+ * @brief Send response to the gate
+ * @param None
+ * @return None
+ ******************************************************************************/
+void LuosBootloader_SendCrc(bootloader_cmd_t response, uint8_t data, uint8_t size)
+{
+    msg_t ready_msg;
+    ready_msg.header.cmd         = BOOTLOADER_RESP;
+    ready_msg.header.target_mode = NODEIDACK;
+    ready_msg.header.target      = 1; // always send to the gate wich launched the detection
```

Comment:
I understand the problem. I connot use routing table API because gate doesn't re-detect the network when it's reboot in bootloader mode. I add a mechanism to get the gate_id from the bootloader without the routing table api.